### DISCUSSION TITLE
Dedicated reference types

### DISCUSSION THREAD

**Discussion by johnwason** [carbon-language/carbon-lang#1717] [Language design]
I am a robotics engineer, and am interested in this project because I find C++ to be very cumbersome compared to more modern alternatives. Unfortunately the more modern alternatives don't have the necessary performance characteristics or the libraries required. 

My suggestion is that a new design include features like dedicated smart reference types and reflection as part of the base specification. .NET C++/CLI used special syntax (various combinations of ^ and %) to represent managed types.  The exact implementation of reference counting and/or garbage collection can be left to a library, but it would be very beneficial if the hooks to implement a smart reference are already present rather than trying to bodge something later like smart pointers and garbage collection in C++.

**Comment by jonmeow** [comment]
I'm not clear what you're trying to suggest here.

Given you express concerns about performance characteristics, please bear in mind that reference counting/GC creates significant overhead. We talk a little about this [in our safety strategy](https://github.com/carbon-language/carbon-lang/blob/trunk/docs/project/principles/safety_strategy.md#guaranteed-memory-safety-programming-models), and [also the FAQ](https://github.com/carbon-language/carbon-lang/blob/trunk/docs/project/faq.md#why-not-a-garbage-collected-language-like-java-kotlin-or-go).

You might consider using Rust if you can: it's a great language and its design addresses a lot of the performance concerns, but at the [cost of a different design requirement](https://github.com/carbon-language/carbon-lang/blob/trunk/docs/project/faq.md#why-not-rust).

**Reply by johnwason** [reply]
Most robotics software have performance requirements that sit between normal applications and hard real-time software. These programs do things like motion planning, computer vision, system level control, etc. They will typically use memory allocation, and are complex enough that this data is passed around to different libraries and threads, so it isn't really possible to know exactly when they can be freed. In a lot of ways this is very similar to game engine requirements, which are using C# more since it is increasingly capable of the required performance, and is a lot easier to work with. The finished software is also generally a lot more reliable since there are less opportunities to make mistakes. C# is something I have been using with robotics, but it has some design aspects like how it tends to allocate on the heap and collect a lot more than is necessary that make it less than ideal.

It is also very desirable to be able to use the same library in different performance regimes. That is why I suggested the ability to vary the exact memory management implementation using a library. In some cases you may be able to use a single threaded reference counting without resorting to atomics, while in others you may need to use a full garbage collector, but the library overall is still unchanged.

Rust is interesting, but almost all of the major robotics libraries are written in C++, and it is unlikely there will be a rewrite of all of these libraries. C# has the same problem of a lack of existing libraries written in C#.

**Reply by YagaoDirac** [reply]
Maybe you need std::pmr(since 17), and all the smart pointers in c++ std library.
GC in c++ is possible but only with libraries.

**Reply by cdmaczane** [reply]
Most larger games are written in C++, not C#. Generally only smaller scale indie games use C# or Unity. The main reason is performance and garbage collection is too much overhead. Garbage collection isn't necessary when you don't create garbage.

**Reply by johnwason** [reply]
@cdmaczane I just checked and Unreal Engine has C++ garbage collection: https://unrealcommunity.wiki/garbage-collection-36d1da I guess there are some games that avoid using reference counting or garbage collection, but they are probably very rare.

**Reply by cdmaczane** [reply]
It is written in C++ not C#. Garbage collection can be a major performance problem in UE4 games, especially on weaker systems such as Switch and mobile phones. Most engines use custom intrusive referencing counting which has some overhead but more predictable performance and lower memory requirements.
